### ISSUE TITLE
Leader not assigned to one of the partitions of a running cluster

### ISSUE THREAD

**Issue by strawhat5** [zeebe-io/zeebe#4893]
**Describe the bug**
We came across an issue where one of the partitions did not have any Leader assigned to it. Due to this many of the requests coming to our Standalone gateway started failing with `DEADLINE_EXCEEDED` error.

**To Reproduce**
There is no specific way to reproduce it. We have a 3 broker, 3 partition cluster (with a separate standalone gateway) which has been running for more than a week. We encountered this issue after one of the broker nodes got restarted yesterday, and post that one partition did not get a Leader assigned.

**Temporary resolution**
By restarting one of the broker nodes, the Leader election algo kicked in, and assigned a Leader to the Follower partition. Does the Leader election only works on restart of a node? 🤔 

**Expected behavior**
All the partitions should have a Leader assigned to them at all times.

**Log/Stacktrace**
<details><summary>Full Stacktrace</summary>
 <p>

```
2020-07-06 11:29:40.694 [GatewayTopologyManager] [gateway-scheduler-zb-actors-0] [36mDEBUG[m io.zeebe.gateway - Received new broker BrokerInfo{nodeId=1, partitionsCount=3, clusterSize=3, replicationFactor=3, partitionRoles={1=FOLLOWER, 2=FOLLOWER, 3=FOLLOWER}, partitionLeaderTerms={}, version=0.23.2}.
2020-07-06 11:29:40.727 [GatewayTopologyManager] [gateway-scheduler-zb-actors-0] [36mDEBUG[m io.zeebe.gateway - Received new broker BrokerInfo{nodeId=2, partitionsCount=3, clusterSize=3, replicationFactor=3, partitionRoles={1=FOLLOWER, 2=FOLLOWER, 3=FOLLOWER}, partitionLeaderTerms={}, version=0.23.2}.
2020-07-06 11:29:40.727 [GatewayTopologyManager] [gateway-scheduler-zb-actors-0] [36mDEBUG[m io.zeebe.gateway - Received new broker BrokerInfo{nodeId=0, partitionsCount=3, clusterSize=3, replicationFactor=3, partitionRoles={1=LEADER, 2=LEADER, 3=FOLLOWER}, partitionLeaderTerms={1=48, 2=32}, version=0.23.2}.
2020-07-06 11:30:04.379 [io.zeebe.gateway.impl.broker.BrokerRequestManager] [gateway-scheduler-zb-actors-1] [31mERROR[m io.zeebe.gateway - Error handling gRPC request
io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: Time out between gateway and broker: Request timed out after PT15S
	at io.grpc.Status.asRuntimeException(Status.java:524) ~[grpc-api-1.28.1.jar:1.28.1]
	at io.zeebe.gateway.EndpointManager.convertThrowable(EndpointManager.java:397) ~[zeebe-gateway-0.23.2.jar:0.23.2]
	at io.zeebe.gateway.impl.job.ActivateJobsHandler.logErrorResponse(ActivateJobsHandler.java:109) ~[zeebe-gateway-0.23.2.jar:0.23.2]
	at io.zeebe.gateway.impl.job.ActivateJobsHandler.lambda$activateJobs$1(ActivateJobsHandler.java:96) ~[zeebe-gateway-0.23.2.jar:0.23.2]
	at io.zeebe.gateway.impl.broker.BrokerRequestManager.lambda$sendRequest$3(BrokerRequestManager.java:148) ~[zeebe-gateway-0.23.2.jar:0.23.2]
	at io.zeebe.gateway.impl.broker.BrokerRequestManager.lambda$sendRequestInternal$5(BrokerRequestManager.java:191) ~[zeebe-gateway-0.23.2.jar:0.23.2]
	at io.zeebe.util.sched.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:33) [zeebe-util-0.23.2.jar:0.23.2]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:76) [zeebe-util-0.23.2.jar:0.23.2]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.23.2.jar:0.23.2]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.23.2.jar:0.23.2]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.23.2.jar:0.23.2]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.23.2.jar:0.23.2]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.23.2.jar:0.23.2]
Caused by: java.util.concurrent.TimeoutException: Request timed out after PT15S
	at io.zeebe.transport.impl.RequestContext.timeout(RequestContext.java:109) ~[zeebe-transport-0.23.2.jar:0.23.2]
	at io.zeebe.transport.impl.AtomixClientTransportAdapter.timeoutFuture(AtomixClientTransportAdapter.java:131) ~[zeebe-transport-0.23.2.jar:0.23.2]
	at io.zeebe.transport.impl.AtomixClientTransportAdapter.lambda$sendRequestWithRetry$0(AtomixClientTransportAdapter.java:60) ~[zeebe-transport-0.23.2.jar:0.23.2]
	... 6 more
2020-07-06 11:30:04.386 [io.zeebe.gateway.impl.broker.BrokerRequestManager] [gateway-scheduler-zb-actors-1] [33mWARN [m io.zeebe.gateway - Failed to activate jobs for type http from partition 3
java.util.concurrent.TimeoutException: Request timed out after PT15S
	at io.zeebe.transport.impl.RequestContext.timeout(RequestContext.java:109) ~[zeebe-transport-0.23.2.jar:0.23.2]
	at io.zeebe.transport.impl.AtomixClientTransportAdapter.timeoutFuture(AtomixClientTransportAdapter.java:131) ~[zeebe-transport-0.23.2.jar:0.23.2]
	at io.zeebe.transport.impl.AtomixClientTransportAdapter.lambda$sendRequestWithRetry$0(AtomixClientTransportAdapter.java:60) ~[zeebe-transport-0.23.2.jar:0.23.2]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:76) [zeebe-util-0.23.2.jar:0.23.2]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.23.2.jar:0.23.2]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.23.2.jar:0.23.2]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.23.2.jar:0.23.2]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.23.2.jar:0.23.2]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.23.2.jar:0.23.2]

```

</p>
</details>

**Environment:**
 - Zeebe Version: `0.23.2`

<details><summary>Configuration</summary>
 <p>

```
Starting standalone gateway with configuration {
  "network" : {
    "host" : "0.0.0.0",
    "port" : 26500,
    "minKeepAliveInterval" : "PT30S"
  },
  "cluster" : {
    "contactPoint" : "prod-zeebe-zeebe-broker:26502",
    "requestTimeout" : "PT15S",
    "clusterName" : "zeebe-cluster",
    "memberId" : "gateway",
    "host" : "100.100.128.4",
    "port" : 26502
  },
  "threads" : {
    "managementThreads" : 2
  },
  "monitoring" : {
    "enabled" : true,
    "host" : "0.0.0.0",
    "port" : 9600
  },
  "security" : {
    "enabled" : false,
    "certificateChainPath" : null,
    "privateKeyPath" : null
  }
}


Starting broker 0 with configuration {
  "network" : {
    "host" : "0.0.0.0",
    "portOffset" : 0,
    "maxMessageSize" : "4MB",
    "advertisedHost" : "prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local",
    "commandApi" : {
      "host" : "0.0.0.0",
      "port" : 26501,
      "advertisedHost" : "prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local",
      "advertisedPort" : 26501,
      "advertisedAddress" : "prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local:26501",
      "address" : "0.0.0.0:26501"
    },
    "internalApi" : {
      "host" : "0.0.0.0",
      "port" : 26502,
      "advertisedHost" : "prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local",
      "advertisedPort" : 26502,
      "advertisedAddress" : "prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local:26502",
      "address" : "0.0.0.0:26502"
    },
    "monitoringApi" : {
      "host" : "0.0.0.0",
      "port" : 9600,
      "advertisedHost" : "prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local",
      "advertisedPort" : 9600,
      "advertisedAddress" : "prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local:9600",
      "address" : "0.0.0.0:9600"
    },
    "maxMessageSizeInBytes" : 4194304
  },
  "cluster" : {
    "initialContactPoints" : [ "prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local:26502", "prod-zeebe-zeebe-broker-1.prod-zeebe-zeebe-broker.prod.svc.cluster.local:26502", "prod-zeebe-zeebe-broker-2.prod-zeebe-zeebe-broker.prod.svc.cluster.local:26502" ],
    "partitionIds" : [ 1, 2, 3 ],
    "nodeId" : 0,
    "partitionsCount" : 3,
    "replicationFactor" : 3,
    "clusterSize" : 3,
    "clusterName" : "zeebe-cluster",
    "gossipFailureTimeout" : 10000,
    "gossipInterval" : 250,
    "gossipProbeInterval" : 1000
  },
  "threads" : {
    "cpuThreadCount" : 2,
    "ioThreadCount" : 2
  },
  "data" : {
    "directories" : [ "/usr/local/zeebe/data" ],
    "logSegmentSize" : "512MB",
    "snapshotPeriod" : "PT15M",
    "logIndexDensity" : 100,
    "logSegmentSizeInBytes" : 536870912,
    "atomixStorageLevel" : "DISK"
  },
  "exporters" : {
    "elasticsearch" : {
      "jarPath" : null,
      "className" : "io.zeebe.exporter.ElasticsearchExporter",
      "args" : {
        "url" : url,
        "bulk" : {
          "delay" : 15,
          "size" : 1000
        },
        "authentication" : {
          "username" : user,
          "password" : password
        },
        "index" : {
          "prefix" : "zeebe-prod",
          "createTemplate" : true,
          "command" : true,
          "event" : true,
          "rejection" : true,
          "deployment" : true,
          "error" : true,
          "incident" : true,
          "job" : true,
          "jobBatch" : true,
          "message" : true,
          "messageSubscription" : true,
          "variable" : true,
          "variableDocument" : true,
          "workflowInstance" : true,
          "workflowInstanceCreation" : true,
          "workflowInstanceSubscription" : true,
          "ignoreVariablesAbove" : 16384
        }
      },
      "external" : false
    },
    "hazelcast" : {
      "jarPath" : "/usr/local/zeebe/exporters/zeebe-hazelcast-exporter.jar",
      "className" : "io.zeebe.hazelcast.exporter.HazelcastExporter",
      "args" : {
        "port" : 5701,
        "enabledValueTypes" : "DEPLOYMENT,JOB,WORKFLOW_INSTANCE,INCIDENT,TIMER,VARIABLE,MESSAGE,MESSAGE_SUBSCRIPTION,MESSAGE_START_EVENT_SUBSCRIPTION",
        "enabledRecordTypes" : "EVENT",
        "name" : "zeebe",
        "capacity" : 100000,
        "timeToLiveInSeconds" : 300,
        "format" : "protobuf"
      },
      "external" : true
    }
  },
  "gateway" : {
    "network" : {
      "host" : "0.0.0.0",
      "port" : 26500,
      "minKeepAliveInterval" : "PT30S"
    },
    "cluster" : {
      "contactPoint" : "0.0.0.0:26502",
      "requestTimeout" : "PT15S",
      "clusterName" : "zeebe-cluster",
      "memberId" : "gateway",
      "host" : "0.0.0.0",
      "port" : 26502
    },
    "threads" : {
      "managementThreads" : 1
    },
    "monitoring" : {
      "enabled" : false,
      "host" : "0.0.0.0",
      "port" : 9600
    },
    "security" : {
      "enabled" : false,
      "certificateChainPath" : null,
      "privateKeyPath" : null
    },
    "enable" : false
  },
  "backpressure" : {
    "enabled" : true,
    "algorithm" : "VEGAS"
  },
  "stepTimeout" : "PT5M"
}

```

</p>
</details>

**Comment by npepinpe** [comment]
Thanks for reporting @strawhat5, we have two different issues related to this: https://github.com/zeebe-io/zeebe/issues/4810 (which was just fixed so will not be in 0.24.0 but rather 0.24.1) and https://github.com/zeebe-io/zeebe/issues/4656 (which is still open as we lacked enough context info to determine and couldn't easily reproduce it).

One thing you could do to help would be to upgrade to 0.24.0 when possible (estimated release date would be sometime next week), where we added more context info about the election process, and set the log level for all Atomix packages to INFO at least - you can do this via the environment variable `ATOMIX_LOG_LEVEL`, or set it specifically for the package, in this case `io.atomix.raft`. After that, if it happens again, the best thing would be to provide us with the logs of each broker.

In the meantime, just a few questions:

1. Only one broker restarted?
1. After it restarted, you say one partition did not get a leader assigned - did the broker which restarted ever finished restarting? If you do `curl -v http://BROKERURL:9600/ready` (e.g. if you're on the broker machine, and do `curl -v http://localhost:9600/ready`, do you get a 204, or a 503?

**Comment by Zelldon** [comment]
Based on the information could also be https://github.com/zeebe-io/zeebe/issues/4557

**Comment by strawhat5** [comment]
Thanks for the quick responses! Great that this is already being tracked in #4810. Sure I will enhance the logging once `0.24.1` gets released.

Regarding your questions @npepinpe :

> Only one broker restarted?

Yes. We run our cluster as a 3 node broker + 2 node standalone gateway cluster, with 3 partitions and replication factor of 3. Out of these nodes, only 1 broker node was restarted and we could trace the start of this issue back to the time when the broker node restarted. After that one of the partitions remained unassigned until we restarted **one other** broker node.

> did the broker which restarted ever finished restarting?

Yes it did. In fact it has been running since yesterday without any further restart, and currently it is also the Leader for 2 of the partitions. This cluster is from our production environment, which is very stable and does not face multiple restarts.

I captured the data of all the broker nodes at the time this issue was occurring. I can provide that in case that can be of help.

@Zelldon This specific issue is not #4557, as I have faced that earlier and that time it got fixed by just restarting the gateway nodes (without touching the broker nodes). But this time restarting the gateway nodes did not fix the issue. I had to restart one of the broker nodes and only after that a Leader was assigned to the 3rd partition.

**Comment by strawhat5** [comment]
One interesting thing that I noted at the time of this issue:

<img width="373" alt="Screenshot 2020-07-06 at 5 08 11 PM" src="https://user-images.githubusercontent.com/10865861/86768727-9d38cb00-c06b-11ea-8960-e1dbbfcd5b4b.png">
Grafana showed that there was a Leader for Raft 1/2/3 (which I am assuming are partitions). Which is incorrect as one of the partitions had a Leader missing. So the Prometheus metrics were also not showing incorrect behaviour at that time. This may be a separate issue though.

**Comment by npepinpe** [comment]
Sorry if I missed this and you already mentioned it, but how did you know there was no leader? Asking since the metrics reported a leader. Keep in mind the metrics is each broker reporting their role for each partition, and it's happened before we had bugs in the gateway where the gateway had a different topology. It can also happen during network partition that brokers incorrectly identify themselves as leader too.

**Comment by strawhat5** [comment]
Let me try to rephrase it. It is my assumption that there was no leader for that partition. And I reached this assumption based on 2 observations:

1. Few weeks back when we faced the gateway topology issue, it was fixed by just restarting the gateway nodes. After the gateway restart, it was able to get the correct topology from the broker.
2. This time when I restarted the gateway nodes, the issue still persisted. Then I decided to restart one of the other broker nodes, for the Leader election algo to be re-triggered (We need more details on when the Leader election gets kicked in or is it always happening like a heartbeat?). After doing so, I could see in the logs that the unassigned partition was allocated a leader. Also on running `zbctl status` on the gateway (without restarting gateway), I was able to get correct topology.

This may not be enough to reach a conclusion, so I can also provide broker data (at the time of occurrence of issue) separately if required, which is quite large to attach here.

**Comment by npepinpe** [comment]
Hey @strawhat5, 0.24.2 was just released with fixes for the issues related to this one - let us know when you get a chance to try it and see if this was indeed the same problem.

**Comment by strawhat5** [comment]
Thanks @npepinpe I will let you know if this issue occurs again. Even with `0.23.4` version we have not seen this issue coming again. It only happened once, but we will keep an eye on it.

You can close this issue if needed, I can reopen in case it happens again.
